core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java

```diff
@@ -117,658 +148,904 @@ public boolean isPrettyPrint() {
-        if (humanReadable) {
-            field(readableFieldName, String.format(Locale.ROOT, "%1.1f%%", percentage));
+    ////////////////////////////////////////////////////////////////////////////
+    // Binary
+    //////////////////////////////////
+
+    public XContentBuilder field(String name, byte[] value) throws IOException {
+        if (value == null) {
+            return nullField(name);
         }
-        field(rawFieldName, percentage);
+        writeBinaryField(name, value);
         return this;
     }
 
-    public XContentBuilder value(Boolean value) throws IOException {
+    public XContentBuilder value(byte[] value) throws IOException {
         if (value == null) {
             return nullValue();
         }
-        return value(value.booleanValue());
-    }
-
-    public XContentBuilder value(boolean value) throws IOException {
-        generator.writeBoolean(value);
+        writeBinary(value);
         return this;
     }
 
-    public XContentBuilder value(ReadableInstant date) throws IOException {
-        return value(date, defaultDatePrinter);
+    public XContentBuilder field(String name, byte[] value, int offset, int length) throws IOException {
+        return field(name).value(value, offset, length);
     }
 
-    public XContentBuilder value(ReadableInstant date, DateTimeFormatter dateTimeFormatter) throws IOException {
-        if (date == null) {
+    public XContentBuilder value(byte[] value, int offset, int length) throws IOException {
+        if (value == null) {
             return nullValue();
         }
-        return value(dateTimeFormatter.print(date));
-    }
-
-    public XContentBuilder value(Date date) throws IOException {
-        return value(date, defaultDatePrinter);
+        writeBinary(value, offset, length);
+        return this;
     }
 
-    public XContentBuilder value(Date date, DateTimeFormatter dateTimeFormatter) throws IOException {
-        if (date == null) {
+    /**
+     * Writes the binary content of the given {@link BytesRef}.
+     *
+     * Use {@link org.elasticsearch.common.xcontent.XContentParser#binaryValue()} to read the value back
+     */
+    public XContentBuilder field(String name, BytesRef value) throws IOException {
+        return field(name).value(value);
+    }
+
+    /**
+     * Writes the binary content of the given {@link BytesRef} as UTF-8 bytes.
+     *
+     * Use {@link XContentParser#utf8Bytes()} to read the value back
+     */
+    public XContentBuilder utf8Field(String name, BytesRef value) throws IOException {
+        return field(name).utf8Value(value);
+    }
+
+    /**
+     * Writes the binary content of the given {@link BytesRef}.
+     *
+     * Use {@link org.elasticsearch.common.xcontent.XContentParser#binaryValue()} to read the value back
+     */
+    public XContentBuilder value(BytesRef value) throws IOException {
```

Comment:
Should this be `binaryValue` so there isn't a `value(BytesRef)` method? Calling this one `value` makes it look like it is the "normal" way to write `BytesRef` but I think actually `utf8Value` is more useful?